support/oidc-discovery-provider/cert_manager_test.go

```diff
@@ -0,0 +1,423 @@
+	})
+
+	t.Run("error when provided cert is invalid", func(t *testing.T) {
+		_, err := NewDiskCertManager(&Config{ServingCertFile: &ServingCertFileConfig{
+			CertFilePath: tmpDir + "/oidcServerCertInvalid.pem",
+			KeyFilePath:  tmpDir + "/oidcServerKey.pem",
+		}}, logger)
+
+		require.EqualError(t, err, "failed to load certificate: tls: failed to find any PEM data in certificate input")
+	})
+
+	t.Run("error when provided key is invalid", func(t *testing.T) {
+		_, err := NewDiskCertManager(&Config{ServingCertFile: &ServingCertFileConfig{
+			CertFilePath: tmpDir + certFilePath,
+			KeyFilePath:  tmpDir + "/oidcServerKeyInvalid.pem",
+		}}, logger)
+
+		require.EqualError(t, err, "failed to load certificate: tls: failed to find any PEM data in key input")
+	})
+
+	t.Run("error when client misses server name", func(t *testing.T) {
+		_, err := tlsConfig.GetCertificate(&tls.ClientHelloInfo{})
+		require.EqualError(t, err, "missing server name")
+	})
+
+	t.Run("error when client send server name with invalid character", func(t *testing.T) {
+		_, err := tlsConfig.GetCertificate(&tls.ClientHelloInfo{
+			ServerName: "example.com:8080",
+		})
+		require.EqualError(t, err, "server name contains invalid character")
+	})
+
+	t.Run("error when client send server name with invalid component count", func(t *testing.T) {
+		_, err := tlsConfig.GetCertificate(&tls.ClientHelloInfo{
+			ServerName: "example",
+		})
+		require.EqualError(t, err, "server name component count invalid")
+	})
+
+	t.Run("error when client send wrong server name", func(t *testing.T) {
+		_, err := tlsConfig.GetCertificate(&tls.ClientHelloInfo{
+			ServerName: "example.com",
+		})
+		require.EqualError(t, err, `server name mismatch: x509: certificate is not valid for any names, but wanted to match example.com`)
+	})
+
+	t.Run("success loading initial certificate from disk", func(t *testing.T) {
+		cert, err := tlsConfig.GetCertificate(chInfo)
+		require.NoError(t, err)
+		require.Len(t, cert.Certificate, 1)
+		x509Cert, err := x509.ParseCertificate(cert.Certificate[0])
+		require.NoError(t, err)
+		require.Equal(t, oidcServerCert, x509Cert)
+	})
+
+	t.Run("success watching cert file changes", func(t *testing.T) {
+		oidcServerCertUpdatedPem := pem.EncodeToMemory(&pem.Block{
+			Type:  "CERTIFICATE",
+			Bytes: oidcServerCertUpdated1.Raw,
+		})
+		err = writeFile(tmpDir+certFilePath, oidcServerCertUpdatedPem)
+		require.NoError(t, err)
+
+		require.Eventuallyf(t, func() bool {
+			cert, err := tlsConfig.GetCertificate(chInfo)
+			if err != nil {
+				return false
+			}
+			require.Len(t, cert.Certificate, 1)
+			x509Cert, err := x509.ParseCertificate(cert.Certificate[0])
+			if err != nil {
+				return false
+			}
+			return reflect.DeepEqual(oidcServerCertUpdated1, x509Cert)
+		}, 1*time.Second, 101*time.Millisecond, "Failed to assert updated certificate")
```

Comment:
Some of these Eventually's might take longer on slow, resource constrained CI/CD environments. I'd bump the timeouts up on these to at least 10 seconds.